### PR TITLE
update notification payload when updating blood donation

### PR DESCRIPTION
**Pull request by mahinuzzaman**
Fixes #
- when updating blood donation, we also need to update the notification payloads we sent to users. here I updated notification payload for the blood donation update request.
- also added some improvements for notification service

- [x] Does not depend on other PRs
- [ ] Depends on other PR being merged first: #

## What kind of change does this PR introduce (check at least one)

- [x] Bugfix
- [x] Improvement
- [ ] New feature
- [x] Refactor
- [ ] Infrastructure
- [ ] Other (please specify)

## How would you describe the changes in a changelog (1-2 lines)?

Provide a brief overview of the changes introduced in this pull request.

## Related Issues/PRs (optional)

Link to the relevant issue(s) or provide a clear description of the problem or feature being addressed.

## How to test

- [ ] Needs regression testing (add QA needed label)

## Checklist

- [x] I have tested these changes locally and they function as expected.
- [ ] I have added or updated necessary documentation to reflect the changes.
- [ ] I have followed the project's coding style and guidelines.
- [ ] I have added appropriate test coverage to ensure the changes are functioning correctly.
- [ ] I have checked for any potential conflicts with other branches or pull requests.
- [ ] I have updated necessary OpenAPI specs

## Reviewer Request (optional)

If you have any specific requests or suggestions for reviewers, mention them here.

## Screenshots (optional)

If applicable, provide any additional screenshots or GIFs that showcase the changes made in this pull request.

## Demo Link (optional)

If applicable, provide a link to a live demo or deployment where the changes can be reviewed.

## Known Issues (optional)

If there are any known issues or limitations with the changes made in this pull request, describe them here.

## Additional Notes (optional)

Add any additional notes or comments you think are relevant to this pull request.

## Deploy risk & follow up

Please assess risk & follow up on deployment of this PR. Elaborate your choice if selecting anything other than 'No risk'.

- [ ] No risk
- [x] Low
- [ ] High

### CODE REVIEW

File: commons/dto/NotificationDTO.ts [hunk 1]
```diff
@@ -18,3 +18,30 @@ export type NotificationDTO = DTO & HasIdentifier & {
...
 }
+
+export type BloodDonationNotificationDTO = DTO & HasIdentifier & {
+  id: string;
+  userId: string;
+  title: string;
+  body: string;
+  type: NotificationType;
+  status: NotificationStatus;
+  payload: {
```
Comment (EbrahimAli-RU) [inline commons/dto/NotificationDTO.ts:29]: Can't we construct payload from existing type?

File: core/application/models/dbModels/NotificationModel.ts [hunk 2]
```diff
@@ -53,8 +38,8 @@ export default class NotificationModel implements NosqlModel<NotificationFields>
     return {
       ...remainingNotificationFields,
       userId,
-      type: parts[1] as NotificationType,
-      id: parts[2]
+      type: parts[0] as NotificationType,
+      id: parts[1]
```
Comment (EbrahimAli-RU) [inline core/application/models/dbModels/NotificationModel.ts:42]: What if parts is an empty array?

File: core/services/aws/notification/sendPushNotification.ts [hunk 3]
```diff
@@ -50,20 +57,63 @@ async function processSQSRecord(record: SQSRecord): Promise<void> {
...
+        requestPostId: body.payload.requestPostId as string,
+        createdAt: body.payload.createdAt as string,
+        bloodQuantity: body.payload.bloodQuantity as string,
+        requestedBloodGroup: body.payload.requestedBloodGroup as string,
+        urgencyLevel: body.payload.urgencyLevel as string,
+        contactNumber: body.payload.contactNumber as string,
+        donationDateTime: body.payload.donationDateTime as string,
+        seekerName: body.payload.seekerName as string,
+        patientName: body.payload.patientName as string,
+        location: body.payload.location as string,
```
Comment (EbrahimAli-RU) [inline core/services/aws/notification/sendPushNotification.ts:93]: Why do we need `as` in every property?

File: core/application/notificationWorkflow/Types.ts [hunk 4]
```diff
@@ -9,15 +9,30 @@ export interface NotificationAttributes {
...
   body: string;
   type: NotificationType;
-  payload: {
-    requestPostId: string;
-    [key: string]: unknown;
-  };
+  payload: BloodDonationPayloadAttributes;
+}
+
+export interface BloodDonationPayloadAttributes {
```
Comment (EbrahimAli-RU) [inline core/application/notificationWorkflow/Types.ts:20]: Same for this

File: core/services/aws/notification/sendPushNotification.ts (hunk 3 above)
Comment (mahinuzzaman) [reply core/services/aws/notification/sendPushNotification.ts:93]: cause, here notification can be multiple types of data. so when its a blood donation notification I had to handle it in a different way.

File: core/application/notificationWorkflow/Types.ts (hunk 4 above)
Comment (mahinuzzaman) [reply core/application/notificationWorkflow/Types.ts:20]: i dont understand, i already separated payload type here

File: commons/dto/NotificationDTO.ts (hunk 1 above)
Comment (mahinuzzaman) [reply commons/dto/NotificationDTO.ts]: Updated in 49f319a8110867cd9a5be505206104734474bfa3

File: core/application/models/dbModels/NotificationModel.ts (hunk 2 above)
Comment (mahinuzzaman) [reply core/application/models/dbModels/NotificationModel.ts]: Updated in 49f319a8110867cd9a5be505206104734474bfa3

File: core/application/notificationWorkflow/Types.ts (hunk 4 above)
Comment (mahinuzzaman) [reply core/application/notificationWorkflow/Types.ts:16]: Updated in 39b4b130f6a0edf11228fe5a7c3faad7d8f4dcf1

File: core/services/aws/notification/sendPushNotification.ts (hunk 3 above)
Comment (mahinuzzaman) [reply core/services/aws/notification/sendPushNotification.ts:93]: Please let me know if there is a better way.

**Review by EbrahimAli-RU — APPROVED** [review]

File: core/application/notificationWorkflow/NotificationService.ts [hunk 5]
```diff
@@ -1,54 +1,122 @@
...
   async publishNotification(
     notificationAttributes: NotificationAttributes,
     userSnsEndpointArn: string,
-    notificationRepository: Repository<NotificationDTO>,
     snsModel: SNSModel
   ): Promise<string> {
     try {
-      const { userId, type, payload } = notificationAttributes
+      await snsModel.publish(notificationAttributes, userSnsEndpointArn)
+      return 'Notified user successfully.'
```
Comment (sakibstark11) [inline core/application/notificationWorkflow/NotificationService.ts:30]: why does need to return a string?

File: core/application/notificationWorkflow/NotificationService.ts [hunk 6]
```diff
@@ -1,54 +1,122 @@
...
+        const updatedNotification: Partial<BloodDonationNotificationDTO> = {
+          id: requestPostId,
+          userId: notification.userId,
+          type: notification.type,
+          payload: { ...notification.payload, ...notificationPayload }
+        }
+
+        await notificationRepository.update(updatedNotification)
+      }
+      return 'Notification updated successfully.'
```
Comment (sakibstark11) [inline core/application/notificationWorkflow/NotificationService.ts:116]: same

File: core/application/notificationWorkflow/NotificationService.ts (hunk 5 above)
Comment (mahinuzzaman) [reply core/application/notificationWorkflow/NotificationService.ts]: Updated in 77a84e14b8a0b4fe53ac3803ea7b36243c3810f7

File: core/application/notificationWorkflow/NotificationService.ts (hunk 6 above)
Comment (mahinuzzaman) [reply core/application/notificationWorkflow/NotificationService.ts]: Updated in 77a84e14b8a0b4fe53ac3803ea7b36243c3810f7

**Review by sakibstark11 — APPROVED** [review]

**Pull request merged**